pageserver/pagebench/src/cmd/basebackup.rs

```diff
@@ -270,3 +268,90 @@ async fn client(
 
     all_work_done_barrier.wait().await;
 }
+
+/// A basebackup client. This allows swithcing out the client protocol implementation.
+#[async_trait]
+trait Client: Send {
+    async fn basebackup(
+        &mut self,
+        ttid: TenantTimelineId,
+        lsn: Option<Lsn>,
+        gzip: bool,
+    ) -> anyhow::Result<Pin<Box<dyn AsyncRead + Send + 'static>>>;
+}
+
+/// A libpq-based Pageserver client.
+struct LibpqClient {
+    inner: pageserver_client::page_service::Client,
+}
+
+impl LibpqClient {
+    async fn new(connstring: &str) -> anyhow::Result<Self> {
+        let inner = pageserver_client::page_service::Client::new(connstring.to_string()).await?;
+        Ok(Self { inner })
+    }
+}
+
+#[async_trait]
+impl Client for LibpqClient {
+    async fn basebackup(
+        &mut self,
+        ttid: TenantTimelineId,
+        lsn: Option<Lsn>,
+        gzip: bool,
+    ) -> anyhow::Result<Pin<Box<dyn AsyncRead + Send + 'static>>> {
+        let req = BasebackupRequest {
+            tenant_id: ttid.tenant_id,
+            timeline_id: ttid.timeline_id,
+            lsn,
+            gzip,
+        };
+        let stream = self.inner.basebackup(&req).await?;
+        Ok(Box::pin(StreamReader::new(
+            stream.map_err(std::io::Error::other),
+        )))
+    }
+}
+
+/// A gRPC client using the raw, no-frills gRPC client.
+struct GrpcClient {
+    inner: page_api::proto::PageServiceClient<tonic::transport::Channel>,
+}
+
+impl GrpcClient {
+    async fn new(connstring: &str) -> anyhow::Result<Self> {
+        let inner = page_api::proto::PageServiceClient::connect(connstring.to_string()).await?;
+        Ok(Self { inner })
+    }
+}
+
+#[async_trait]
+impl Client for GrpcClient {
+    async fn basebackup(
+        &mut self,
+        ttid: TenantTimelineId,
+        lsn: Option<Lsn>,
+        gzip: bool,
+    ) -> anyhow::Result<Pin<Box<dyn AsyncRead + Send + 'static>>> {
+        // TODO: support gzip compression.
+        ensure!(!gzip, "gRPC client does not support gzip compression yet");
```

Comment:
what would it take to support gzip?

Reply:
Not much, probably an hour or two of work. I plan to get to it in the next couple of days, it's blocked on some other pending PRs.